SEP-994: Shared Communication Practices/Guidelines (#1002)
modelcontextprotocol/modelcontextprotocol: docs/community/communication.mdx

Reviewer: We should call out our Code of Conduct and link to it.

## Communication Channels

We support three primary communication channels - the public Discord server, [GitHub Issues](https://github.com/modelcontextprotocol/modelcontextprotocol/issues), and [GitHub Discussions](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions) in the [main project repository](https://github.com/modelcontextprotocol/modelcontextprotocol).

### Discord

For real-time contributor discussion, support, and collaboration. The server is designed around **MCP contributors** and is not intended
to be the place for general MCP support.

The Discord server will have both public and private channels.

#### Public Channels (Default)

- **Purpose**: Open community engagement, collaborative development, and transparent project coordination.
- Primary use cases:
  - **Public SDK and tooling development**: All development, from ideation to release planning, happens in public channels (e.g., #typescript-sdk-dev, #inspector-dev). This replaces the previous hybrid model (public/private channel combo).
  - **Working and interest group discussions** (`#client-implementors`, `#agents-wg`, etc.)

>     - **Working Group**: Some specific goal or project in mind (such as an SDK, inspector, registry, server-identity, load-balancing, etc).

Author: @tadasant - want to take a pen on that once we have this SEP merged? Would love you to bring your expertise to shaping that part of the community.

    - **Interest Group**: An abstract gathering of folks that might raise a range of various topics. Some might get actioned on as one-offs, others might spin into Working Groups.
  - **Community onboarding**, help channels, and contribution guidance.
  - **Community feedback** and collaborative brainstorming.
  - Public **office hours** and **maintainer availability**.

#### Private channels (Exceptions)

- **Purpose**: Confidential coordination and sensitive matters that cannot be discussed publicly. Access will be restricted to designated maintainers.
- **Strict criteria for private use**:
  - **Security incidents** (CVEs, protocol vulnerabilities).
  - **People matters** (maintainer-related discussions, code of conduct policies).
  - Select channels will be configured to be **read-only**. This can be good for example for maintainer decision making.
  - Coordination requiring **immediate** or otherwise **focused response** with a limited audience
- **Transparency requirements**:

>   - **All decisions** affecting the community **must be documented** in GitHub Discussions and/or Issues.

Author: Love this @ihrpr - will add!

  - Coordination requiring **immediate** or otherwise **focused response** with a limited audience
- **Transparency requirements**:
  - **All decisions** affecting the community **must be documented** in GitHub Discussions and/or Issues.
  - Private channels are to be used as **temporary "incident rooms,"** not for routine development.

Any significant discussion on Discord that leads to a potential decision or proposal must be moved to a GitHub Discussion or GitHub Issue to create a persistent, searchable record. Proposals will then be promoted to full-fledged PRs with associated work items (GitHub Issues) as needed.

### GitHub Discussions

For structured, long-form updates on project direction, features, and community topics.

When to use:

- Formal governance decisions requiring community input and voting
- Project roadmap planning and milestone discussions
- Announcements and release communications
- Community polls and consensus-building processes

### GitHub Issues

For bug reports, feature tracking, and actionable development tasks.

When to use:

- Submit SEP proposals (following the [SEP guidelines](./sep-guidelines))

> - Bug reports with reproducible steps

Reviewer 2: ```suggestion
Feature requests with context and rationale
```

Reviewer 3: In my opinion, pre-SEP feature requests are better suited as GitHub Discussions instead of GitHub Issues.

In open source projects, feature requests tend to accumulate because it is difficult to definitively say "no" to a feature.

I think we should avoid accumulating open Issues whenever possible, whereas I think it's fine to accumulate Discussions.

Reviewer 3: The difference between Issues and Discussions is that Issues are a form of To-Do list.  A large number of open (unresolved) GitHub Issues is often seen as an indicator of poor project health.

More to your point, there is asymmetric dynamic at play.  Anyone in the world can propose something, and small handful of people must address all of those proposals.  Spending more time triaging is great, but it isn't a solution.  By moving the proposals out of the To-Do list, it puts some of the responsibility back onto the submitter and the community to raise awareness and flesh out ideas before demanding the time and attention of maintainers.

Reviewer 4: Yeah that def makes sense, maybe I'm misusing "triage", I kind of had in mind something like very low-level "Issue maintainers" let's say, who could properly label things, find duplicates, press people to either reply or close the Issue, track Issues that are more well fleshed out, etc. My question would be what is the path from Discussions to Issues? Partly this is because it seems to me that many Discussion posts get very little engagement, so you could say that would then fall on the community to be more vigilant, I'm not sure, maybe that's a good approach.

Reviewer 4: I'll also admit I don't have much experience with large OSS projects like this, so my POV is kind of an outsider who's been making efforts to contribute according to SoP, and observing the pain points.

Reviewer 3: > Problem with feature requests starting in Discussions means that maintainers have one more place that they have to troll through.

I'm not sure I understand.  By default, when you watch a GitHub repo, you will get notifications for both Issues and Discussions.  (Note that I am advocating using repo-level Discussions, not org-level Discussions.  See also https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1002#discussion_r2228691855.)

> Also, most people out in the world just open an issue on OSS projects when they want a feature. They may not have read the bible we're building about appropriate interaction with the project.

First of all, I think we should be encouraging more thoughtful participation.  But also, it only takes a few clicks to convert an Issue into a Discussion.  If someone opens an Issue and sees that a moderator has converted it to a Discussion, they will likely take the hint.  (And if not, we can point to the guidelines.)

Reviewer 5: Can we potentially take this on a per-repo basis?

For example, I've quite liked keeping the modelcontextprotocol/registry's Issues list clean and only things we actually intend to implement. Ideas have been fielded in Discussions.

OTOH, projects like Inspector don't have Discussions activated. And maybe that's fine. In the absence of Discussions, everything just goes in Issues, and the maintainer team can triage them as they please.

So the overall guidance could be, "when there is a Discussions option on a repo, split Discussions & Issues this way; in the absence of a Discussions board, just do it all in Issues".

Reviewer 2: > First of all, I think we should be encouraging more thoughtful participation. 

That can easily happen in an issue that proposes a feature request

> But also, it only takes a few clicks to convert an Issue into a Discussion. If someone opens an Issue and sees that a moderator has converted it to a Discussion, they will likely take the hint. (And if not, we can point to the guidelines.)

IMHO, having to clean up the issue list by moving feature requests to discussions is needless friction. Nothing is gained but thickness of rules. People doing what they naturally do in open source projects is easy to handle in the usual way, no one needs to take any hints. 

> OTOH, projects like Inspector don't have Discussions activated. And maybe that's fine. In the absence of Discussions, everything just goes in Issues, and the maintainer team can triage them as they please.

Happy for the discussion -> issue protocol to be in place for the spec, but glad it can be simpler and more straightforward for the project repos. There, potential contributors usually cruise the issues list for feature requests that they might want to implement, but they may not go off into the discussions if they were present. We could attempt to "educate" them at every turn, but it seems easier to just follow the natural way of the open source world and not try and impose a more complicated structure than necessary.

- Documentation improvements with specific scope
- CI/CD problems and infrastructure issues
- Release tasks and milestone tracking

### Security Issues

**Do not post security issues publicly.** Instead:

> 1. Use the private security reporting process. For protocol-level security issues, send a note to `security@modelcontextprotocol.io` with the details.

Reviewer 3: ```suggestion
1. Use the private security reporting process. For protocol-level security issues, send a note to [security@modelcontextprotocol.io](mailto:security@modelcontextprotocol.io) with the details.
```

2. Contact lead and/or core maintainers directly.
3. Follow responsible disclosure guidelines.

## Decision Records

All MCP decisions are documented and captured in public channels.

> - **Technical decisions**. [GitHub Issues](https://github.com/modelcontextprotocol/modelcontextprotocol/issues) and SEPs. Specification changes published [on the Model Context Protocol website](https://modelcontextprotocol.io/introduction).
> - **Process changes**. [Community documentation.](https://modelcontextprotocol.io/community/governance)
> - **Governance decisions and updates**. [GitHub Discussions](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions).

Reviewer 3: ```suggestion
- **Technical decisions**: [GitHub Issues](https://github.com/modelcontextprotocol/modelcontextprotocol/issues) and SEPs. 
- **Specification changes**: [On the Model Context Protocol website](https://modelcontextprotocol.io/specification/draft/changelog).
- **Process changes**: [Community documentation](https://modelcontextprotocol.io/community/governance).
- **Governance decisions and updates**: [GitHub Discussions](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions).
```

When documenting decisions, we will retain as much context as possible:

> - Decision makers.
> - Background context and motivation.
> - Options that were considered.
> - Rationale for the chosen approach.
> - Implementation steps.

Reviewer 3: ```suggestion
- Decision makers
- Background context and motivation
- Options that were considered
- Rationale for the chosen approach
- Implementation steps
```

  - **People matters** (maintainer-related discussions, code of conduct policies).
  - Select channels will be configured to be **read-only**. This can be good for example for maintainer decision making.
  - Coordination requiring **immediate** or otherwise **focused response** with a limited audience
- **Transparency requirements**:
  - **All decisions** affecting the community **must be documented** in GitHub Discussions and/or Issues, and will be labeled with `notes`.
  - Private channels are to be used as **temporary "incident rooms,"** not for routine development.

Any significant discussion on Discord that leads to a potential decision or proposal must be moved to a GitHub Discussion or GitHub Issue to create a persistent, searchable record. Proposals will then be promoted to full-fledged PRs with associated work items (GitHub Issues) as needed.

### GitHub Discussions

For structured, long-form updates on project direction, features, and community topics.

When to use:

- Formal governance decisions requiring community input and voting
- Project roadmap planning and milestone discussions
- Announcements and release communications

> - Community polls and consensus-building processes

Reviewer 3: Pursuant to https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1002#discussion_r2228794078.

```suggestion
- Community polls and consensus-building processes
- Feature requests with context and rationale
  - If a particular repository does not have GitHub Discussions enabled, feel free to open a GitHub Issue instead.
```

### GitHub Issues

For bug reports, feature tracking, and actionable development tasks.

When to use:

- Submit SEP proposals (following the [SEP guidelines](./sep-guidelines))
- Bug reports with reproducible steps

> - Feature requests with context and rationale

Reviewer 3: Pursuant to https://github.com/modelcontextprotocol/modelcontextprotocol/pull/1002#discussion_r2228794078.

```suggestion
```

---
title: Contributor Communication
description: Communication strategy and framework for the Model Context Protocol community
---

> This document provides practical guidance for communicating and collaborating within the Model Context Protocol (MCP) project.
> It outlines the communication channels, workflows, and processes used by the MCP community.

Reviewer 5: ```suggestion
This document provides practical guidance for communicating and collaborating within the Model Context Protocol (MCP) project. It outlines the communication channels, workflows, and processes used by the MCP community.
```

[formatting nit]

## Communication Channels

> We support three primary communication channels: the public Discord server, [GitHub Issues](https://github.com/modelcontextprotocol/modelcontextprotocol/issues), and [GitHub Discussions](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions) in the [main project repository](https://github.com/modelcontextprotocol/modelcontextprotocol).

Reviewer 5: A thought, just to make sure this is the intention (not blocking merging this PR): can we turn off/deprecate the org-level discussions [here](https://github.com/orgs/modelcontextprotocol/discussions)? 

@dsp-ant has indicated previously that he likes how it filters noise (more serious suggestions get made in the spec repo-level Discussions), but I wonder if @jonathanhefner 's work to help organize things on the Issues/Discussions front has addressed that to some degree (+ the SEP process is now where things lift above the noise). Curious if @jonathanhefner agrees with deprecating that, and @dsp-ant would be OK with it now.

cc @olaservo who I believe raised that conversation previously

Reviewer: Just let me know when you want me to do that.

Reviewer 3: I don't want to disrupt ongoing conversations, so I will start transferring recently active and popular (highly upvoted or commented) discussions to the `modelcontextprotocol/modelcontextprotocol` repo.  I don't have permissions to transfer discussions to the SDK repos, so we will need other maintainers to transfer SDK-related discussions.

After the discussions are transferred, we can make an Announcement post and lock those forums to maintainers only.  (I think the lock will apply to discussion threads as well, but I am not sure.)

Reviewer 3: I have transferred most recently active / popular discussions from the org to this repo.

I did not transfer any of the [show and tell](https://github.com/orgs/modelcontextprotocol/discussions/categories/show-and-tell) discussions.  To me, those don't quite fit in this specific repo, but, if it's something we want, I can transfer the most recent threads.

Regarding SDK-related discussions, the only two notably active discussions I found were related to the Java SDK: https://github.com/orgs/modelcontextprotocol/discussions/246 and https://github.com/orgs/modelcontextprotocol/discussions/322.  Perhaps someone with sufficient permissions could transfer those to the Java SDK repo?

### Discord

> For real-time contributor discussion and collaboration. The server is designed around **MCP contributors** and is not intended
> to be the place for general MCP support.

Reviewer 5: ```suggestion
For real-time contributor discussion and collaboration. The server is designed around **MCP contributors** and is not intended to be the place for general MCP support.
```

[formatting nit]

The Discord server will have both public and private channels.

#### Public Channels (Default)

- **Purpose**: Open community engagement, collaborative development, and transparent project coordination.
- Primary use cases:

>   - **Public SDK and tooling development**: All development, from ideation to release planning, happens in public channels (e.g., `#typescript-sdk-dev`, `#inspector-dev`). This replaces the previous hybrid model (public/private channel combo).

Reviewer: ```suggestion
  - **Public SDK and tooling development**: All development, from ideation to release planning, happens in public channels (e.g., `#typescript-sdk-dev`, `#inspector-dev`).
```

  - **Working and interest group discussions** (`#client-implementors`, `#agents-wg`, etc.)
    - **Working Group**: Some specific goal or project in mind (such as an SDK, inspector, registry, server-identity, load-balancing, etc).
    - **Interest Group**: An abstract gathering of folks that might raise a range of various topics. Some might get actioned on as one-offs, others might spin into Working Groups.
  - **Community onboarding** and contribution guidance.
  - **Community feedback** and collaborative brainstorming.
  - Public **office hours** and **maintainer availability**.
- Unapproved uses:

>   - MCP user support: participants are expected to to read official documentation and start new GitHub Discussions for questions or support.

Reviewer 6: ```suggestion
  - MCP user support: participants are expected to read official documentation and start new GitHub Discussions for questions or support.
```

  - Service or product marketing: interactions on this Discord are expected to be vendor-neutral and not used for brand-building or sales. Mentions of brands or products are discouraged outside of being used as examples or responses to conversations that start off focused on the specification.

#### Private channels (Exceptions)

- **Purpose**: Confidential coordination and sensitive matters that cannot be discussed publicly. Access will be restricted to designated maintainers.
- **Strict criteria for private use**:
  - **Security incidents** (CVEs, protocol vulnerabilities).
  - **People matters** (maintainer-related discussions, code of conduct policies).
  - Select channels will be configured to be **read-only**. This can be good for example for maintainer decision making.

>   - Coordination requiring **immediate** or otherwise **focused response** with a limited audience

Reviewer: ```suggestion
  - Coordination requiring **immediate** or otherwise **focused response** with a limited audience.
```

> - **Transparency requirements**:

Reviewer: ```suggestion
- **Transparency**:
```

>   - **All decisions** affecting the community **must be documented** in GitHub Discussions and/or Issues, and will be labeled with `notes`.

Reviewer: We should be careful around this wording. We might want to make decisions about individuals that we cannot or don't want to make public at all.

  - Private channels are to be used as **temporary "incident rooms,"** not for routine development.

Any significant discussion on Discord that leads to a potential decision or proposal must be moved to a GitHub Discussion or GitHub Issue to create a persistent, searchable record. Proposals will then be promoted to full-fledged PRs with associated work items (GitHub Issues) as needed.

### GitHub Discussions

For structured, long-form updates on project direction, features, and community topics.

When to use:

> - Formal governance decisions requiring community input and voting

Reviewer: This feels more like a SEP and would go into issues?!

---
title: Contributor Communication
description: Communication strategy and framework for the Model Context Protocol community
---

This document provides practical guidance for communicating and collaborating within the Model Context Protocol (MCP) project. It outlines the communication channels, workflows, and processes used by the MCP community.

All communication within the MCP community is governed by our [Code of Conduct](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/main/CODE_OF_CONDUCT.md). We expect all participants to maintain respectful, professional, and inclusive interactions across all channels.

## Communication Channels

> We support three primary communication channels: the public Discord server, [GitHub Issues](https://github.com/modelcontextprotocol/modelcontextprotocol/issues), and [GitHub Discussions](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions) in the [main project repository](https://github.com/modelcontextprotocol/modelcontextprotocol).

Reviewer 7: there's this one from the CWG github:
https://github.com/modelcontextprotocol-community/working-groups/blob/main/README.md?plain=1#L22

`https://discord.gg/6CSzBmMkjX`

> We support three primary communication channels: the public Discord server, [GitHub Issues](https://github.com/modelcontextprotocol/modelcontextprotocol/issues), and [GitHub Discussions](https://github.com/modelcontextprotocol/modelcontextprotocol/discussions) in the [main project repository](https://github.com/modelcontextprotocol/modelcontextprotocol).

Reviewer 5: Agree that we can just keep using `https://discord.gg/6CSzBmMkjX`

  - **Working and interest group discussions** (`#client-implementors`, `#agents-wg`, etc.)
    - **Working Group**: Some specific goal or project in mind (such as an SDK, inspector, registry, server-identity, load-balancing, etc).
    - **Interest Group**: An abstract gathering of folks that might raise a range of various topics. Some might get actioned on as one-offs, others might spin into Working Groups.
  - **Community onboarding** and contribution guidance.
  - **Community feedback** and collaborative brainstorming.
  - Public **office hours** and **maintainer availability**.
- Avoid:
  - MCP user support: participants are expected to read official documentation and start new GitHub Discussions for questions or support.
  - Service or product marketing: interactions on this Discord are expected to be vendor-neutral and not used for brand-building or sales. Mentions of brands or products are discouraged outside of being used as examples or responses to conversations that start off focused on the specification.

#### Private channels (Exceptions)

> 

Reviewer 8: How can people locate moderators that could create these private channels?

### GitHub Discussions

For structured, long-form updates on project direction, features, and community topics.

When to use:

- Project roadmap planning and milestone discussions
- Announcements and release communications
- Community polls and consensus-building processes
- Feature requests with context and rationale
  - If a particular repository does not have GitHub Discussions enabled, feel free to open a GitHub Issue instead.

### GitHub Issues

For bug reports, feature tracking, and actionable development tasks.

When to use:

- Submit SEP proposals (following the [SEP guidelines](./sep-guidelines))
- Bug reports with reproducible steps
- Documentation improvements with specific scope
- CI/CD problems and infrastructure issues
- Release tasks and milestone tracking

### Security Issues

**Do not post security issues publicly.** Instead:

1. Use the private security reporting process. For protocol-level security issues, send a note to [security@modelcontextprotocol.io](mailto:security@modelcontextprotocol.io) with the details.

> 2. Contact lead and/or core maintainers directly.

Reviewer 8: ```suggestion
2. Contact lead and/or [core maintainers](https://modelcontextprotocol.io/community/governance#current-core-maintainers) directly.
```
